client/pooling.zig

```diff
@@ -0,0 +1,179 @@
+            }
+
+            if (self.size == other.size) {
+                return .eq; // equals means it's an exact match
+            }
+
+            if (self.size < other.size) {
+                return .gt; // greater than means it's a match, but not a perfect match
+            }
+
+            unreachable;
+        }
+    };
+}
+
+pub const StringKey = struct {
+    value: []const u8,
+
+    pub fn compare(self: StringKey, other: StringKey) std.math.Order {
+        return if (std.mem.eql(u8, self.value, other.value)) .eq else .lt;
+    }
+};
+
+pub fn FrameReferencedResourcePool(comptime Context: type, comptime Key: type, comptime Value: type, comptime create_val: anytype, comptime release_val: anytype, comptime frames_to_keep_entry: comptime_int) type {
+    // errors to make this very clear on what to do
+    comptime {
+        if (std.meta.activeTag(@typeInfo(Key)) != .@"struct")
+            @compileError("Key type must be a struct. There are examples of key types in this source file.");
+
+        if (!@hasDecl(Key, "compare"))
+            @compileError("Key type is missing an 'eql' function. There are examples of key types in this source file.");
+
+        if (!@hasField(Key, "value"))
+            @compileError("Key type is missing the 'value' field. There are examples of key types in this source file.");
+    }
+
+    return struct {
+        pub const Entry = struct {
+            frames_since_usage: u32,
+            value: Value,
+            key: Key,
+        };
+
+        const Self = @This();
+
+        lock: std.Thread.Mutex,
+        // FIXME: this should be a doubly linked list for performance sake
+        entries: std.ArrayListUnmanaged(Entry),
+        context: Context,
+
+        pub fn init(context: Context) Self {
+            return .{
+                .lock = .{},
+                .entries = .empty,
+                .context = context,
+            };
+        }
+
+        pub fn acquire(self: *Self, key: Key) !Entry {
+            self.lock.lock();
+            defer self.lock.unlock();
+
+            const none_found = std.math.maxInt(usize);
+
+            var smallest_index: usize = none_found;
+            var smallest_size: usize = none_found;
+
+            for (self.entries.items, 0..) |entry, i| {
+                const comparison = entry.key.compare(key);
+
+                if (comparison == .lt)
+                    continue;
+
+                if (comparison != .gt) {
+                    smallest_index = i;
+                    if (@hasField(Key, "size"))
+                        smallest_size = entry.key.size;
+                }
+
+                if (comparison == .eq)
+                    break;
+            }
+
+            return if (smallest_index == none_found) .{
+                .key = key,
+                .frames_since_usage = 0,
+                .value = try create_val(self.context, key),
+            } else self.entries.swapRemove(smallest_index);
```

Comment:
We can make this logic a lot more readable like this.

```suggestion
            var best_index: usize = none_found;
			var best_size: usize = none_found;

            for (self.entries.items, 0..) |entry, i| {
                switch(entry.key.compare(key)) {
                	.lt => continue,
                	.gt => {
                		if (@hasField(Key, "size")) {
                			// If this is sized, only update the best index if the new entry's size is less than the current best size
                			if (entry.key.size < best_size) {
                				best_index = i;
                				best_size = @intCast(entry.key.size);
                			}
                		} else {                	
                			best_index = i;
                		}
              		},
                	.eq => {
                		best_index = i;
                		break;
              		},
                }
            }

            return if (best_index == none_found) .{
                .key = key,
                .frames_since_usage = 0,
                .value = try create_val(self.context, key),
            } else self.entries.swapRemove(smallest_index);
```

Reply:
I disagree that the switch case makes this more readable... it feels about the same. I like the new variable naming though.

The new condition with sizing also doesn't match assumptions, and causes a buffer that isn't big enough to be chosen: `vkCmdCopyBufferToImage(): pRegions[0] is trying to copy 16384 bytes to/from the (VkBuffer 0x16c300000016c3[Pooled Transfer Buffer (size 64)]) which exceeds the VkBuffer total size of 64 bytes.`

I have no earthly clue why. This function is really hard to parse.

Reply:
Could you push what you have so far? I'll take a look.

Reply:
Fixed, key comparison operands were swapped